### DISCUSSION TITLE
Swapping retract speeds might be confusing in hindsight (#722)

### DISCUSSION THREAD

**Discussion by JohnScottUK** [sn4k3/UVtools#735] [General]
In hindsight, I'm not sure swapping retract[0] & retract[1] values when handling Anycubic M3 Premium files was a good idea.

Appreciate it was instigated by me. However, it might be a little confusing.

When I look at settings created from Anycubic, I know that the order is (unusually) Lift[0] -> Lift[1] -> Retract[1] -> Retract[0]

What I think you have done (please correct if I am misunderstanding) is map the retract[0] into 2) retract and retract[1] into 1) retract settings in UVTools when loading PM3R files (swapping them over).

I get this is to keep UVTools implying the order is lift[0]->lift[1]->retract[0]->retract[1]

However, What people see in Anycubic Photon Workshop and then see in UVTools might be confusing because they are numbered but are no longer mapped in order ([0] mapped into Retract and [1] mapped into 2) retract).

![image](https://github.com/sn4k3/UVtools/assets/68079471/2e8e2fad-cbc3-4b41-b13f-ab2e0b0c88fb)

The cause of my initial confusion was the tooltip for 2) Retract says "Slow" which it probably is for all other printers/formats, but not PM3R on M3 Premium file formats. Maybe if the labels were not numbers, or the tooltip changed, that would help.

Appologies for causing work. Thought I would open a discussion rather than bug etc so you can decide what you want to do.

**Comment by sn4k3** [comment]
Nothing changed in UVtools logic.
It was indeed a bug where I thought that anycubic retract 0/1 would match UVtools retract 1/2, but in fact they use reverse logic. 
The only alteration was by reading the anycubic file and associate retract0 with UVtools retract2 and retract 1 with UVTools retract, as so the UVtools logic has not changed and your assumption is correct.

UVtools still uses: Lift 1 -> Lift2 -> Retract 1 -> Retract 2. Which IMO is the most correct (Ordinal Events) and it's also what all other slicers that I'm aware uses beside Anycubic. For that reason, I'm not changing that logic.

I do agree that the terms and how layout is presented can be confusing and hard to edit, preferably the terms 1/2 should not exists and the terms: Slow/Fast should be used. However, that presents another "problem", in UVtools for printers that lacks of TSMC the lift would be presented as "Slow Lift" and retract as "Fast Retract" raising again confusion. Even if TSMC is present user may opt to not use it (0) and then raise again confusion if fast or slow is the first. First lift would show as slow and retract as fast, what field should I use now if I don't want to use TSMC? This is like a dead-loop for UVtools that is format not-aware and must go with anything.

UVtools was designed pior to TSMC exists, where there weren't additional lifts and then adapted with those terms (Which I still think it's the best var representation), that put the code logic under control / understandable as it doesn't care about slow/fast but rather who comes first and after. 

Also, the level of abstraction used in the dialog make it not possible to do much better.
That tool layout and structure needs a revamp, in future I need redo it for the sake of simplicity, which would follow something similar to what chitubox uses (Fields side by side) making it easier to understand and spare space. However, expect that first shown retract is the fast one, just like Chitubox and others present.

Until the dialog gets revamp, you are stuck with the mental exercise of swap those on the softwares, after some time it should be more automatic...

BTW, what is your use case for the editing? Is not so often required to edit parameters as they can be set from slicer.

**Reply by JohnScottUK** [reply]
I understand what you've done and am happy with it. Lychee has swapped the retract speeds and so have you now. Lychee doesn't number them where as Anycubic and UVtools do.

I was concerned the swap might trip Anycubic Workshop users and didn't want your tool to be criticised because of my bug report.

I'm still trying Anycubic/Lychee/Chitubox slicers which show the 4 values differently.

I think the problem might be Chitubox because I used its defaults for my printer which when observed retract slowly then accelerate towards the bed (I think they need to swap the figures when displaying like you now & Lychee). I will raise a big report on them if I can find how to do that...

As to my use case. I was using UVtools to convert PruSlicer SL to PM3R when trying it out. But I also use it to check for islands, resin traps overhangs etc when sliced which it's a great tool for.

**Reply by JohnScottUK** [reply]
Maybe calling them
1st Lift
2nd Lift
1st Retract
2nd Retract

Just a thought but happy with it now I understand Chitubox looks the wrong way round.

**Reply by sn4k3** [reply]
> I'm still trying Anycubic/Lychee/Chitubox slicers which show the 4 values differently.

Yes, each software has their way to present and naming. But looks like they keep the order of events (1,2,3,4) [Slow-Fast-Fast-Slow], only anycubic does not (1,2,4,3) [Slow-Fast-Slow-Fast] but of obvious this is just a software arrangement because machine will do [Slow-Fast-Fast-Slow] after all if the speeds are well assigned on their respective fields.

For example both lychee and chitubox are in agree with the order

![image](https://github.com/sn4k3/UVtools/assets/113281/fa5a8704-9633-4298-8acc-253b6cc2c054)
![image](https://github.com/sn4k3/UVtools/assets/113281/96fb5c95-34c5-490a-b5b1-5a1bc795cc34)

Generally, Anycubic is the worse software from the 3, I would use Lychee or Tango, PrusaSlicer or even the chitubox, either is way better than anycubic.

> I think the problem might be Chitubox because I used its defaults for my printer which when observed retract slowly then accelerate towards the bed

They might have same bug as UVtools and Lychee had for anycubic machines, again no one predicted the swap convention from anycubic, they use reverse logic in software which makes no sense from a natural perspective. So, I not be amazed if chitubox also has/had that "bug". Did you try their latest version?

> As to my use case. I was using UVtools to convert PruSlicer SL to PM3R when trying it out. But I also use it to check for islands, resin traps overhangs etc when sliced which it's a great tool for.

I mean for the Tool - Edit Parameters as it looks like you use it more often...

> Maybe calling them
1st Lift
2nd Lift
1st Retract
2nd Retract

Definitely labels can be changed. I use "2) ..." as a way to instant identify the TSMC parameters without pior read.
But anyway I expect to revamp the dialog

**Reply by sn4k3** [reply]
Check the new version

**Reply by JohnScottUK** [reply]
Just had a quick look. Excellent dialog window now. Clear and consistent with other tools (Anycubic take note :-) )
